Fix #166 by removing GET BSO limit

mostlygeek opened the pull request:
- There should be no limit on the amount of records a client
  can fetch in one request. This confirmed by this [1] comment
  on the python sync server
- Remove configuration option for limiting GET BSOs
- Refactor code for checking validity of limit and offset
- Made default behaviour return all requested BSOs

[1] https://github.com/mozilla-services/server-syncstorage/blob/b8feef17147af8819d50c9fae459e3e5de149d42/syncstorage/views/__init__.py#L247-L250

mostlygeek committed:
Fix #166 by removing GET BSO limit

- There should be no limit on the amount of records a client
  can fetch in one request. This confirmed by this [1] comment
  on the python sync server
- Remove configuration option for limiting GET BSOs
- Refactor code for checking validity of limit and offset
- Made default behaviour return all requested BSOs

[1] https://github.com/mozilla-services/server-syncstorage/blob/b8feef17147af8819d50c9fae459e3e5de149d42/syncstorage/views/__init__.py#L247-L250

rfk commented on syncstorage/db.go:
```diff
@@ -887,12 +887,15 @@ func (d *DB) getBSOs(
+
+		if offset != 0 {
+			limitStmt += " OFFSET ?"
+			values = append(values, offset)
```
The protocol permits you to specify `offset` without `limit`, e.g. you could say "give me the first 100 items" and then use a second request to say "give me the rest from this offset".  I don't believe any clients actually do this, and I can't think of a reason why they would, but it's permitted by the protocol.

rfk reviewed (approved):
One nit w.r.t. protocol semantics, but otherwise this looks good to me.

mostlygeek committed:
Always provide a LIMIT clause for getting BSO

- use -1 for unbounded SELECT queries. Allows use of the OFFSET
  clause without limiting the number of records returned

mostlygeek commented:
@rfk I refactored the code so 

- a LIMIT clause is always used
- unbounded queries use `-1` for the limit value since OFFSET requires a limit

please r? again

rfk reviewed (approved):
👍

rfk commented:
Well except for the build bustage obviously ;-)

mostlygeek committed:
Fix failure of pysync functional test test_get_collection

The change to the limit logic broke pysync's test_get_collection test
since -1 is now an appropriate limit value internally. This commit
changes the behaviour of DB.getBSOs(...) so the check for additional
records only happens when a LIMIT > 0 is request.

- refactor code so offset is calculated correctly
- slight optimization as the SELECT for a count of all records *only*
  happens when limit is > 0 (only for bounded queries)
- removed total int from GetResults struct (it was wrong anyways)
- removed tests for GetResults total value as it can be calculated

mostlygeek commented:
@rfk one more r? please. I fixed the broken build but wound up ripping out some more bad/old logic.

rfk commented on syncstorage/db.go:
```diff
@@ -928,15 +920,29 @@ func (d *DB) getBSOs(
+		countQuery := "SELECT COUNT(1) NumRows FROM BSO " + where + " " + orderBy
+		if err := tx.QueryRow(countQuery, values...).Scan(&totalRows); err != nil {
+			return nil, err
+
```
extra newlines here?

mostlygeek replied:
i blame gofmt. :)

rfk commented on syncstorage/db.go:
```diff
@@ -928,15 +920,29 @@ func (d *DB) getBSOs(
+	// the upper bound and return the pertinent information
+	if limit > 0 {
+		var totalRows int
+		countQuery := "SELECT COUNT(1) NumRows FROM BSO " + where + " " + orderBy
```
This could be an expensive extra query for large collections; a future optimization could be to use `LIMIT {{limit}} + 1` in your original query, and then set `more = true` if you managed to read more than `{{limit}}` items.  NBD right now though.

mostlygeek replied:
that's smart! 👍

mostlygeek committed:
begone extra newline!
